vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -30,361 +35,315 @@
+         box_id < device_ctx.connection_boxes.num_connection_box_types();
+         box_id++) {
+        auto& matrix = cost_map_[iseg][box_id];
+        if (matrix.dim_size(0) == 0 || matrix.dim_size(1) == 0) {
+            printf("cost EMPTY for box_id = %lu\n", box_id);
+            continue;
+        }
+        printf("cost for box_id = %lu\n", box_id);
+        double sum = 0.0;
+        for (unsigned iy = 0; iy < matrix.dim_size(1); iy++) {
+            for (unsigned ix = 0; ix < matrix.dim_size(0); ix++) {
+                const auto& entry = matrix[ix][iy];
+                if (entry.valid()) {
+                    sum += entry.delay;
+                }
             }
-
-            int node_cost_ind = device_ctx.rr_nodes[node_ind].cost_index();
-            int node_seg_ind = device_ctx.rr_indexed_data[node_cost_ind].seg_index;
-            if (node_seg_ind == seg_index_) {
-                index_ += 1;
-                return node_ind;
+        }
+        double avg = sum / ((double)matrix.dim_size(0) * (double)matrix.dim_size(1));
+        for (unsigned iy = 0; iy < matrix.dim_size(1); iy++) {
+            for (unsigned ix = 0; ix < matrix.dim_size(0); ix++) {
+                const auto& entry = matrix[ix][iy];
+                if (entry.valid() && entry.delay > avg) {
+                    printf("o");
+                } else {
+                    printf(".");
+                }
             }
+            printf("\n");
         }
-
-        return UNDEFINED;
     }
+}
 
-  private:
-    int start_x_;
-    int start_y_;
-    t_rr_type rr_type_;
-    int seg_index_;
-    size_t index_;
-};
-
-// Minimum size of search for channels to profile.  kMinProfile results
-// in searching x = [0, kMinProfile], and y = [0, kMinProfile[.
-//
-// Making this value larger will increase the sample size, but also the runtime
-// to produce the lookahead.
-static constexpr int kMinProfile = 1;
-
-// Maximum size of search for channels to profile.  Once search is outside of
-// kMinProfile distance, lookahead will stop searching once:
-//  - At least one channel has been profiled
-//  - kMaxProfile is exceeded.
-static constexpr int kMaxProfile = 7;
-
-static int search_at(int iseg, int start_x, int start_y, t_routing_cost_map* cost_map) {
-    const auto& device_ctx = g_vpr_ctx.device();
+std::list<std::pair<int, int> > CostMap::list_empty() const {
+    std::list<std::pair<int, int> > results;
+    for (int iseg = 0; iseg < (int)cost_map_.dim_size(0); iseg++) {
+        for (int box_id = 0; box_id < (int)cost_map_.dim_size(1); box_id++) {
+            auto& matrix = cost_map_[iseg][box_id];
+            if (matrix.dim_size(0) == 0 || matrix.dim_size(1) == 0) results.push_back(std::make_pair(iseg, box_id));
+        }
+    }
+    return results;
+}
 
-    int count = 0;
-    int dx = 0;
-    int dy = 0;
+static void assign_min_entry(Cost_Entry& dst, const Cost_Entry& src) {
+    if (src.delay < dst.delay) dst.delay = src.delay;
```

Comment:
These values should move together?  Take the min of either seems invalid.  Can you justify why you did this?

Reply:
It's calculating the lower bound for both values.

Reply:
This still needs to be changed.